include/deal.II/base/symbolic_function.templates.h

```diff
@@ -0,0 +1,227 @@
+  {
+    return time;
+  }
+
+
+
+  template <int dim, typename RangeNumberType>
+  const std::vector<Differentiation::SD::Expression> &
+  SymbolicFunction<dim, RangeNumberType>::get_symbolic_function_expressions()
+    const
+  {
+    return function;
+  }
+
+
+
+  template <int dim, typename RangeNumberType>
+  SymbolicFunction<dim, RangeNumberType>
+  SymbolicFunction<dim, RangeNumberType>::time_derivative() const
+  {
+    std::vector<Differentiation::SD::Expression> df_dt(function.size());
+    for (unsigned int i = 0; i < function.size(); ++i)
+      df_dt[i] = function[i].differentiate(time);
+    return SymbolicFunction<dim, RangeNumberType>(df_dt, argument, time);
+  }
+
+
+
+  template <int dim, typename RangeNumberType>
+  RangeNumberType
+  SymbolicFunction<dim, RangeNumberType>::value(
+    const Point<dim> & p,
+    const unsigned int component) const
+  {
+    AssertIndexRange(component, function.size());
+    return Differentiation::SD::substitute_and_evaluate<RangeNumberType>(
+      function[component], substitute(p));
+  }
+
+
+
+  template <int dim, typename RangeNumberType>
+  Tensor<1, dim, RangeNumberType>
+  SymbolicFunction<dim, RangeNumberType>::gradient(
+    const Point<dim> & p,
+    const unsigned int component) const
+  {
+    AssertIndexRange(component, function.size());
+    return Differentiation::SD::substitute_and_evaluate<RangeNumberType>(
+      function_gradient[component], substitute(p));
+  }
+
+
+
+  template <int dim, typename RangeNumberType>
+  RangeNumberType
+  SymbolicFunction<dim, RangeNumberType>::laplacian(
+    const Point<dim> & p,
+    const unsigned int component) const
+  {
+    AssertIndexRange(component, function.size());
+    return Differentiation::SD::substitute_and_evaluate<RangeNumberType>(
+      function_laplacian[component], substitute(p));
+  }
+
+
+
+  template <int dim, typename RangeNumberType>
+  SymmetricTensor<2, dim, RangeNumberType>
+  SymbolicFunction<dim, RangeNumberType>::hessian(
+    const Point<dim> & p,
+    const unsigned int component) const
+  {
+    AssertIndexRange(component, function.size());
+    return SymmetricTensor<2, dim, RangeNumberType>(
+      Differentiation::SD::substitute_and_evaluate<RangeNumberType>(
+        function_hessian[component], substitute(p)));
+  }
+
+
+
+  template <int dim, typename RangeNumberType>
+  template <typename StreamType>
+  StreamType &
+  SymbolicFunction<dim, RangeNumberType>::print(StreamType &out) const
+  {
+    for (unsigned int i = 0; i < dim; ++i)
+      out << argument[i] << ", ";
+    out << time << " -> " << function[0];
+    for (unsigned int i = 1; i < function.size(); ++i)
+      out << "; " << function[i];
+    return out;
+  }
+
+
+
+  template <int dim, typename RangeNumberType>
+  Differentiation::SD::types::substitution_map
+  SymbolicFunction<dim, RangeNumberType>::substitute(const Point<dim> &p) const
+  {
+    return Differentiation::SD::make_substitution_map(
```

Comment:
If my suggestion about caching the user-given substitution map is followed through on, then I would suggest wrapping this in a call to `resolve_explicit_dependencies()`.

Reply:
That should not be necessary. The way I implemented your suggestion, is a hybrid version:

I keep a user substitution map around, which can be updated by the user. Whenever that happens, a mutable function (different from the one specified by the user), is generated, together with all its derivatives. If the user updates the map, then a new function is generated using the original function (which is left untouched from construction time), and the new substitution map.

This leaves the user with a consistent function, where you can replace parameters, and that it is only updated when necessary (i.e., I do not need to resolve all explicit dependencies at evaluation time, but just once, when the user supplies a substitution map, or at construction time).

Reply:
Yeah, you're right. I'm overly defensive about this. Let's leave the responsibility of making this call to the user.